raids: add world region to scout overlay

melkypie opened the pull request:
Adds the region of the world the scout was performed in.

Before: W303         Username
Now:    W303 (GER)   Username

TheJets commented:
This should ideally say if it's an east coast world or west coast @melkypie Pretty important for ping

Nightfirecat commented on runelite-client/src/main/java/net/runelite/client/plugins/raids/RaidsOverlay.java:
```diff
@@ -113,7 +119,20 @@ public Dimension render(Graphics2D graphics)
+			else
+			{
+				worldString = "W" + client.getWorld();
+			}
```
Rather than effectively writing the code for the start of the world string twice, append the region string on a successful world service lookup.
```suggestion
			String worldString = "W" + client.getWorld();

			WorldResult worldResult = worldService.getWorlds();
			if (worldResult != null)
			{
				String worldRegion = getRegion(currentWorld.getLocation());
				worldString += " (" + worldRegion + ')';
			}
```

Nightfirecat commented on runelite-client/src/main/java/net/runelite/client/plugins/raids/RaidsOverlay.java:
```diff
@@ -188,4 +207,19 @@ else if (plugin.getRoomBlacklist().contains(room.getName().toLowerCase()))
 		return panelComponent.render(graphics);
 	}
+
+	private String getRegion(int locationId)
```
This should be defined in an enum somewhere.

Nightfirecat commented on runelite-client/src/main/java/net/runelite/client/plugins/raids/RaidsOverlay.java:
```diff
@@ -188,4 +207,19 @@ else if (plugin.getRoomBlacklist().contains(room.getName().toLowerCase()))
+				return "UK";
+			case 3:
+				return "AUS";
+			default:
```
What number is this default case? This seems very unsafe in the event new regions are added.

melkypie commented:
Added the `WorldRegion `enum to `net.runelite.http.api.worlds` and changed some other things that were relying on the old location id. Also made the requested changes

Nightfirecat commented on http-api/src/main/java/net/runelite/http/api/worlds/WorldRegion.java:
```diff
@@ -0,0 +1,70 @@
+	GERMANY(7, "DE", "DEU");
+
+	private final int id;
+	private final String twoLetterCode;
```
Instead of adding a comment at the top of the enum's body explaining the fields, add javadoc to the fields needing explaining. (renaming the field would make that clearer as well)
```suggestion
	/**
	 * ISO-3166-1 alpha-2 country code
	 */
	private final String alpha2;
```

Nightfirecat commented on http-api/src/main/java/net/runelite/http/api/worlds/WorldRegion.java:
```diff
@@ -0,0 +1,70 @@
+
+	private final int id;
+	private final String twoLetterCode;
+	private final String threeLetterCode;
```
Same here
```suggestion
	/**
	 * ISO-3166-1 alpha-3 country code
	 */
	private final String alpha3;
```

Nightfirecat commented on runelite-client/src/main/java/net/runelite/client/plugins/worldhopper/WorldTableRow.java:
```diff
@@ -326,26 +313,12 @@ private JPanel buildWorldField()
 
-		JLabel flag = new JLabel(getFlag(world.getLocation()));
+		JLabel flag = new JLabel(new ImageIcon(ImageUtil.getResourceStreamFromClass(WorldHopperPlugin.class,
+			"flag_" + world.getLocation().getTwoLetterCode().toLowerCase() + ".png")));
```
This change would duplicate the icon/image initialization per table row generated. Instead, cache the icons in a `static` block (like was done before) using a `Map<WorldRegion, ImageIcon>`.

Nightfirecat commented on http-api/src/main/java/net/runelite/http/api/worlds/WorldRegion.java:
```diff
@@ -0,0 +1,70 @@
+	 * @param locationId the location id of world
+	 * @return WorldRegion the region of the world
+	 */
+	public static WorldRegion valueOf(int locationId)
```
Mark this `@Nullable`
```suggestion
	@Nullable
	public static WorldRegion valueOf(int locationId)
```

melkypie replied:
`@Nullable` doesn't seem to be available in http-api as travis was throwing out errors that it couldn't find `com.sun.istack.internal`

Nightfirecat commented on runelite-client/src/main/java/net/runelite/client/plugins/worldhopper/WorldTableRow.java:
```diff
@@ -64,12 +62,17 @@
 	private static final Color FREE_WORLD = new Color(200, 200, 200);
 	private static final Color LEAGUE_WORLD = new Color(157, 237, 1);
 
+	private static final Map<WorldRegion, ImageIcon> flags;
```
This should follow constant naming conventions.
```suggestion
	private static final Map<WorldRegion, ImageIcon> FLAGS;

```

melkypie replied:
Fixed

Adam- commented on runelite-client/src/main/java/net/runelite/client/plugins/worldhopper/WorldTableRow.java:
```diff
@@ -64,12 +62,17 @@
-		FLAG_US = new ImageIcon(ImageUtil.getResourceStreamFromClass(WorldHopperPlugin.class, "flag_us.png"));
-		FLAG_GER = new ImageIcon(ImageUtil.getResourceStreamFromClass(WorldHopperPlugin.class, "flag_ger.png"));
+		ImmutableMap.Builder<WorldRegion, ImageIcon> builder = new ImmutableMap.Builder<>();
+		for (WorldRegion worldRegion : WorldRegion.values())
```
This type of thing becomes easy to miss later when adding a new world region. Keeping the previous code and changing getFlag to `getFlag(WorldRegion region)` can compile-time error due to the switch if you cover all possible enum values.

melkypie replied:
Wouldn't it still throw an error since it couldn't find the file, if only an enum entry was added?

Adam- replied:
If it was run yes, but I don't think this is even tested so we wouldn't notice.

melkypie replied:
Added a test that would check for this

Adam- commented:
> 
> 
> This should ideally say if it's an east coast world or west coast @melkypie Pretty important for ping

We don't have a good way to determine this currently
